mlflow/pyfunc/scoring_server/__init__.py

```diff
@@ -132,6 +137,74 @@ def infer_and_parse_json_input(json_input, schema: Schema = None):
         )
 
 
+def _decode_json_input(json_input):
+    """
+    :param json_input: A JSON-formatted string representation of TF serving input or a Pandas
+                       DataFrame, or a stream containing such a string representation.
+    :return: A dictionary representation of the JSON input.
+    """
+    if isinstance(json_input, dict):
+        decoded_input = json_input
+    else:
+        try:
+            decoded_input = json.loads(json_input)
+        except json.decoder.JSONDecodeError as ex:
+            raise MlflowException(
+                message=(
+                    "Failed to parse input from JSON. Ensure that input is a valid JSON"
+                    f" formatted string. Error: '{ex}'. Input: \n{json_input}\n"
+                ),
+                error_code=BAD_REQUEST,
+            )
+
+    if isinstance(decoded_input, dict):
+        return decoded_input
+    if isinstance(decoded_input, list):
+        message = "Received a list"
+        raise MlflowException(
+            message=f"{REQUIRED_INPUT_FORMAT}. {message}. {SCORING_PROTOCOL_CHANGE_INFO}",
+            error_code=BAD_REQUEST,
+        )
+    else:
+        message = f"Received unexpected input type '{type(decoded_input)}'"
+        raise MlflowException(
+            message=f"{REQUIRED_INPUT_FORMAT}. {message}.", error_code=BAD_REQUEST
+        )
+
+
+def _split_data_and_params(json_input):
+    input_dict = _decode_json_input(json_input)
+    data = {k: v for k, v in input_dict.items() if k in SUPPORTED_FORMATS}
+    params = input_dict.get("params")
+    return data, params
+
+
+def infer_and_parse_data(data, schema: Schema = None):
+    """
+    :param data: A dictionary representation of TF serving input or a Pandas
+                 DataFrame, or a stream containing such a string representation.
+    :param schema: Optional schema specification to be used during parsing.
+    """
+
+    format_keys = set(data.keys()).intersection(SUPPORTED_FORMATS)
+    if len(format_keys) != 1:
+        message = f"Received dictionary with input fields: {list(data.keys())}"
+        raise MlflowException(
+            message=f"{REQUIRED_INPUT_FORMAT}. {message}. {SCORING_PROTOCOL_CHANGE_INFO}",
+            error_code=BAD_REQUEST,
+        )
+    input_format = format_keys.pop()
+    if input_format in (INSTANCES, INPUTS):
+        return parse_tf_serving_input(data, schema=schema)
+
+    elif input_format in (DF_SPLIT, DF_RECORDS):
+        # NB: skip the dataframe_ prefix
+        pandas_orient = input_format[10:]
+        return dataframe_from_parsed_json(
+            data[input_format], pandas_orient=pandas_orient, schema=schema
+        )
```

Comment:
Since these are not connected by an `else` resolution, could we separate these conditional statements?

```suggestion
    if input_format in (DF_SPLIT, DF_RECORDS):
        pandas_orient = input_format[10:]  # skip the dataframe_ prefix
        return dataframe_from_parsed_json(
            data[input_format], pandas_orient=pandas_orient, schema=schema
        )

    raise MlflowException(
        message=(
            f"Unsupported input format: {input_format}. "
            f"Expected one of {SUPPORTED_FORMATS}"
        ),
        error_code=BAD_REQUEST,
    )
```

Reply:
We don't need to raise extra MlflowException at the end because at the beginning of this function we check 'if len(format_keys) != 1:' and throw an exception there :D